// devexpress-dashboard/viewer-parts/viewer/_events-helper.js
/**
* DevExpress Dashboard (_events-helper.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
require("devextreme/ui/widget/ui.widget");
var _item_data_tuple_1 = require("../../data/item-data/_item-data-tuple");
var chart_item_1 = require("../../model/items/chart/chart-item");
var data_dashboard_item_1 = require("../../model/items/data-dashboard-item");
var gauge_item_1 = require("../../model/items/gauge/gauge-item");
var grid_item_1 = require("../../model/items/grid/grid-item");
var item_widget_event_args_1 = require("./item-widget-event-args");
exports.createItemClickEventArgs = function (dashboardItem, dataPoint, requestUnderlyingDataFunc) {
    var itemData = dashboardItem._getItemData();
    var patchAxisName = function (axisName) {
        if (axisName === undefined)
            axisName = 'Default';
        return axisName;
    };
    var getAxis = function (axisName) {
        return itemData.getAxis(patchAxisName(axisName));
    };
    var getAxisPoint = function (axisName) {
        axisName = patchAxisName(axisName);
        return getAxis(axisName).getPointByUniqueValues(dataPoint.getValues(axisName));
    };
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getData: function () {
            return itemData;
        },
        getAxisPoint: getAxisPoint,
        getMeasures: function () {
            return itemData.getMeasuresByIds(dataPoint.getMeasureIds());
        },
        getDeltas: function () {
            var ids = dataPoint.getDeltaIds(), deltas = [];
            ids.forEach(function (id) {
                deltas.push(itemData.getDeltaById(id));
            });
            return deltas;
        },
        getDimensions: function (axisName) {
            return getAxis(axisName).getDimensions();
        },
        requestUnderlyingData: function (onCompleted, dataMembers) {
            var axisPoints = [];
            itemData.getAxisNames().forEach(function (axisName) {
                axisPoints.push(getAxisPoint(axisName));
            });
            requestUnderlyingDataFunc(dashboardItem.componentName(), {
                axisPoints: axisPoints,
                dataMembers: dataMembers
            }, onCompleted);
        }
    };
};
exports.createItemSelectionChangedEventArgs = function (dashboardItem, tuples) {
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getCurrentSelection: function () {
            var axisPointTuples = [];
            tuples.forEach(function (tuple) {
                axisPointTuples.push(dashboardItem._getItemData().createTuple(tuple));
            });
            return axisPointTuples;
        }
    };
};
exports.createItemElementCustomColorEventArgs = function (dashboardItem, eventArgs) {
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getTargetElement: function () {
            return new _item_data_tuple_1.itemDataTuple(eventArgs.targetElement);
        },
        getMeasures: function () {
            return dashboardItem._getItemData().getMeasuresByIds(eventArgs.measureIds);
        },
        getColor: function () {
            return eventArgs.color;
        },
        setColor: function (color) {
            eventArgs.color = color;
        }
    };
};
exports.createItemInteractivityEventArgs = function (dashboardItem, interactivityOptions) {
    return {
        itemName: dashboardItem.componentName(),
        dashboardItem: dashboardItem,
        getSelectionMode: function () { return interactivityOptions.selectionMode; },
        setSelectionMode: function (value) { interactivityOptions.selectionMode = value; },
        isHighlightingEnabled: function () { return interactivityOptions.hoverEnabled; },
        enableHighlighting: function (value) { interactivityOptions.hoverEnabled = value; },
        getTargetAxes: function () { return interactivityOptions.targetAxes; },
        setTargetAxes: function (value) { interactivityOptions.targetAxes = value; },
        getDefaultSelection: function () {
            var realTuples = [];
            interactivityOptions.defaultSelectedValues.forEach(function (tuple) {
                realTuples.push(dashboardItem._getItemData().createTuple(tuple));
            });
            return realTuples;
        },
        setDefaultSelection: function (realTuples) {
            var tuples = [], tuple;
            realTuples.forEach(function (realTuple) {
                tuple = [];
                interactivityOptions.targetAxes.forEach(function (axisName) {
                    tuple.push({
                        AxisName: axisName,
                        Value: realTuple.getAxisPoint(axisName).getUniquePath()
                    });
                });
                tuples.push(tuple);
            });
            interactivityOptions.defaultSelectedValues = tuples;
        }
    };
};
var createBaseWidgetEventArgs = function (dashboardItem) {
    var itemData = dashboardItem instanceof data_dashboard_item_1.DataDashboardItem ? dashboardItem._getItemData() : null;
    return {
        itemName: dashboardItem.componentName(),
        itemData: itemData,
        dashboardItem: dashboardItem,
        chartContext: dashboardItem instanceof chart_item_1.ChartItem ? new item_widget_event_args_1.ChartContext(dashboardItem) : null,
        gridContext: dashboardItem instanceof grid_item_1.GridItem ? new item_widget_event_args_1.GridContext(dashboardItem) : null,
        gaugeContext: dashboardItem instanceof gauge_item_1.GaugeItem ? new item_widget_event_args_1.GaugeContext(dashboardItem) : null
    };
};
exports.createWidgetEventArgs = function (dashboardItem, widget) {
    return __assign({ getWidget: function () { return widget; } }, createBaseWidgetEventArgs(dashboardItem));
};
exports.createWidgetOptionsEventArgs = function (dashboardItem, options) {
    return __assign({ options: options }, createBaseWidgetEventArgs(dashboardItem));
};
