// devexpress-dashboard/designer/extensions/interactivity-panel.js
/**
* DevExpress Dashboard (interactivity-panel.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var control_options_1 = require("../../common/control-options");
var _dashboard_item_menu_1 = require("../items/_dashboard-item-menu");
var _interactivity_properties_composer_1 = require("../items/_interactivity-properties-composer");
var name = 'item-interactivity-panel';
var InteractivityPanelExtension = (function () {
    function InteractivityPanelExtension(dashboardControl) {
        this.dashboardControl = dashboardControl;
        this.name = name;
        this._subscriptions = [];
    }
    InteractivityPanelExtension.prototype._contextMenuSubscriber = function (itemContextMenu) {
        if (!!itemContextMenu) {
            var item = this.dashboardControl._dashboardSurface().selectedDashboardItem();
            this._updateContextMenu(itemContextMenu, item);
        }
    };
    InteractivityPanelExtension.prototype._updateContextMenu = function (menu, item) {
        if (item._isInteractivityAllowed()) {
            if (menu instanceof _dashboard_item_menu_1.ItemMenuViewModel) {
                menu.contextMenuItems.push({
                    menuItemId: this.name,
                    icon: 'dx-dashboard-interactivity',
                    title: 'DashboardWebStringId.Interactivity',
                    panelWidth: _dashboard_item_menu_1.DashboardItemMenuSizes.OptionsPanelWidth,
                    templateName: 'dx-dashboard-options',
                    detailVisible: ko.observable(false),
                    showMenu: function () {
                        var composer = new _interactivity_properties_composer_1.InteractivityPropertiesComposer(menu.propertiesController);
                        var tabs = composer.composeTabs(item);
                        menu.propertiesController.mainModel({
                            data: {
                                model: item,
                                propertiesTabs: ko.observableArray(tabs),
                            }
                        });
                    },
                    customData: {
                        propertiesController: menu.propertiesController,
                    },
                    index: 200
                });
            }
        }
    };
    InteractivityPanelExtension.prototype.start = function () {
        var contextMenuExtension = this.dashboardControl.findExtension('item-menu');
        if (contextMenuExtension) {
            this._subscriptions.push(contextMenuExtension._itemContextMenu.subscribe(this._contextMenuSubscriber, this));
            this._contextMenuSubscriber(contextMenuExtension._itemContextMenu());
        }
    };
    InteractivityPanelExtension.prototype.stop = function () {
        this._subscriptions.forEach(function (s) { return s.dispose(); });
        this._subscriptions = [];
    };
    return InteractivityPanelExtension;
}());
exports.InteractivityPanelExtension = InteractivityPanelExtension;
control_options_1.designerExtensions[name] = function (dashboardControl, options) { return new InteractivityPanelExtension(dashboardControl); };
