// devexpress-dashboard/viewer-parts/viewer-items/_widget-viewer-item.js
/**
* DevExpress Dashboard (_widget-viewer-item.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var _utils_1 = require("../../data/_utils");
var _widgets_viewer_1 = require("../widgets/widgets-viewer/_widgets-viewer");
var _base_item_1 = require("./_base-item");
var widgetViewerItem = (function (_super) {
    __extends(widgetViewerItem, _super);
    function widgetViewerItem(container, options) {
        return _super.call(this, container, options) || this;
    }
    widgetViewerItem.prototype._clearSelectionUnsafe = function () {
        this.widgetsViewer.clearSelections();
    };
    widgetViewerItem.prototype.getInfoUnsafe = function () {
        return _utils_1.deepExtend(_super.prototype.getInfoUnsafe.call(this), this.widgetsViewer.getSizeParams());
    };
    widgetViewerItem.prototype.renderContentUnsafe = function (element, changeExisting, afterRenderCallback) {
        var that = this, options = that._getWidgetViewerOptions(), isInAsyncRendering = true;
        options.viewer.onAllItemsRenderComplete = function () {
            if (that.widgetsViewer) {
                afterRenderCallback();
            }
            else {
                isInAsyncRendering = false;
            }
        };
        if (changeExisting && that.widgetsViewer) {
            that.widgetsViewer.option(options);
        }
        else {
            that.widgetsViewer = that.createWidgetViewer(element, options);
        }
        return isInAsyncRendering;
    };
    widgetViewerItem.prototype.createWidgetViewer = function (element, options) {
        return new _widgets_viewer_1.dxWidgetsViewer(element, options);
    };
    widgetViewerItem.prototype._getContainerPositionUnsafe = function () {
        var that = this, position = _super.prototype._getContainerPositionUnsafe.call(this), itemInfo = that.getInfo(), scrollSize = itemInfo && itemInfo.scroll && itemInfo.scroll.vertical ? itemInfo.scroll.size : 0;
        position.offsetX -= scrollSize;
        return position;
    };
    widgetViewerItem.prototype._getSpecificWidgetViewerOptions = function () {
        var _this = this;
        return {
            itemOptions: {
                encodeHtml: this._isEncodeHtml(),
                itemWidgetOptionsPrepared: function (options) { return _this._raiseItemWidgetOptionsPrepared(options); }
            }
        };
    };
    widgetViewerItem.prototype._getWidgetType = function () {
        return;
    };
    widgetViewerItem.prototype._isHoverEnabled = function () {
        return this._selectionMode() !== 'none';
    };
    widgetViewerItem.prototype._configureHover = function (selectionValues) {
        var hoverEnabled = selectionValues !== null && this._isHoverEnabled() && _utils_1.allowSelectValue(selectionValues);
        return {
            hoverEnabled: hoverEnabled,
            cursor: hoverEnabled ? 'pointer' : 'default'
        };
    };
    widgetViewerItem.prototype._getWidgetViewerOptions = function () {
        var _this = this;
        var viewModel = this.options.ViewModel, contentDescription = viewModel ? viewModel.ContentDescription : undefined, commonOptions = {
            viewer: {
                redrawOnResize: false
            }
        };
        if (this.dataController)
            this.dataController.setSourceItemProperties = function (sourceItem, element, properties) { return _this._setSourceItemProperties(sourceItem, element, properties); };
        commonOptions.dataSource = this._getDataSource();
        commonOptions.viewer.onclick = this._getOnClickHandler();
        commonOptions.viewer.onhover = this._getOnHoverHandler();
        commonOptions.viewer.widgetType = this._getWidgetType();
        commonOptions.viewer.method = contentDescription ? this._convertContentArrangementMode(contentDescription.ArrangementMode) : 'auto';
        commonOptions.viewer.count = contentDescription ? contentDescription.LineCount : 1;
        commonOptions.viewer.supportAnimation = this._supportAnimation();
        var resultOptions = _utils_1.deepExtend(commonOptions, this._getSpecificWidgetViewerOptions());
        this._ensureOptions(resultOptions);
        return resultOptions;
    };
    widgetViewerItem.prototype._supportAnimation = function () {
        return false;
    };
    widgetViewerItem.prototype._getDataSource = function () {
        if (this.dataController)
            return this.dataController.getDataSource();
    };
    widgetViewerItem.prototype._getElementInteractionValue = function (element, viewModel) {
        return element.tag;
    };
    widgetViewerItem.prototype._getOnClickHandler = function () {
        var that = this;
        return function (e) {
            that._raiseItemClick(e.item);
        };
    };
    widgetViewerItem.prototype._getOnHoverHandler = function () {
        var that = this;
        return function (e) {
            that._raiseItemHover(e.item, e.state);
        };
    };
    widgetViewerItem.prototype._convertContentArrangementMode = function (contentArrangementMode) {
        switch (contentArrangementMode) {
            case 'FixedColumnCount':
                return 'column';
            case 'FixedRowCount':
                return 'row';
            default:
                return 'auto';
        }
    };
    widgetViewerItem.prototype._resizeUnsafe = function () {
        _super.prototype._resizeUnsafe.call(this);
        this.widgetsViewer.redraw();
    };
    widgetViewerItem.prototype.updateContentStateUnsafe = function () {
        var _this = this;
        this.widgetsViewer.itemsList.forEach(function (viewer) {
            viewer.setHoverEnabledState(_this._getCustomHoverEnabled());
        });
    };
    widgetViewerItem.prototype._setSourceItemProperties = function (sourceItem, elementModel, props) {
    };
    widgetViewerItem.prototype._isMultiDataSupported = function () {
        return true;
    };
    widgetViewerItem.prototype._ensureOptions = function (options) {
    };
    return widgetViewerItem;
}(_base_item_1.baseItem));
exports.widgetViewerItem = widgetViewerItem;
