// devexpress-dashboard/model/items/options/metadata/_dashboard-item-coloring-options.js
/**
* DevExpress Dashboard (_dashboard-item-coloring-options.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _base_metadata_1 = require("../../../metadata/_base-metadata");
exports.useGlobalColors = {
    propertyName: 'useGlobalColors', modelName: '@UseGlobalColors', displayName: 'DashboardWebStringId.Colorization.ColorSchemeType', defaultVal: true, simpleFormAdapterItem: 'buttonGroupEditor', from: _base_metadata_1.parseBool,
    valuesArray: [{
            value: true,
            displayValue: 'DashboardWebStringId.Coloring.Global'
        }, {
            value: false,
            displayValue: 'DashboardWebStringId.Coloring.Local'
        }],
    category: _base_metadata_1.PropertyCategory.Coloring
};
exports.measuresColoringMode = {
    propertyName: 'measuresColoringMode', modelName: '@MeasuresColoringMode', displayName: 'DashboardWebStringId.Coloring.MeasuresColoringMode', defaultVal: 'Default', simpleFormAdapterItem: 'buttonGroupEditor',
    values: {
        'Default': 'DashboardWebStringId.ColoringModeAuto',
        'None': 'DashboardWebStringId.ColoringModeOff',
        'Hue': 'DashboardWebStringId.ColoringModeOn'
    }
};
exports.dashboardItemColoringOptionsSerializationsInfo = [exports.useGlobalColors, exports.measuresColoringMode];
