// devexpress-dashboard/common/notification-controller/_notificator-view-model.js
/**
* DevExpress Dashboard (_notificator-view-model.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var NotificationControllerViewModel = (function () {
    function NotificationControllerViewModel() {
        var _this = this;
        this._visible = false;
        this.visible = ko.observable(false);
        this.type = ko.observable('');
        this.notifications = ko.observableArray([]);
        this.suspended = ko.observable(false);
        this.displayTime = ko.computed(function () {
            if (_this.type() !== 'success') {
                return 60 * 1000;
            }
            return 10 * 1000;
        });
        this.onInitialized = function (args) {
            _this._widget = args.component;
        };
    }
    NotificationControllerViewModel.prototype.reset = function () {
        this.visible(false);
        this._visible = false;
    };
    NotificationControllerViewModel.prototype.updateNotification = function (type, title, detail) {
        var _this = this;
        var notification = { title: title, detail: detail || null };
        if (this.visible() && this.type() === type) {
            this.notifications.push(notification);
            if (this._widget)
                this._widget.repaint();
        }
        else {
            var isFirstLoading = !this.visible();
            this.visible(false);
            this.notifications.removeAll();
            this.notifications.push(notification);
            this.type(type);
            this._visible = true;
            if (isFirstLoading) {
                setTimeout(function () { _this.visible(_this._visible); }, 300);
            }
            else {
                this.visible(this._visible);
            }
        }
    };
    return NotificationControllerViewModel;
}());
exports.NotificationControllerViewModel = NotificationControllerViewModel;
