// devexpress-dashboard/model/format-rules/metadata/_format-rules-common.js
/**
* DevExpress Dashboard (_format-rules-common.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var format_condition_bar_options_1 = require("../conditions/format-condition-bar-options");
var appearance_settings_1 = require("../style-settings/appearance-settings");
var bar_style_settings_1 = require("../style-settings/bar-style-settings");
var icon_settings_1 = require("../style-settings/icon-settings");
exports.negativeInfinity = '-Infinity';
exports.fieldTypes = {
    Text: 'System.String',
    DateTime: 'System.DateTime',
    Bool: 'System.Boolean',
    Integer: 'System.Int32',
    Float: 'System.Single',
    Double: 'System.Double',
    Decimal: 'System.Decimal',
    Enum: 'Enum',
    Custom: 'Custom',
    Unknown: 'Unknown'
};
exports.complexValueType = { propertyName: 'type', modelName: '@Type', defaultVal: null };
exports.complexValueValue = { propertyName: 'value', modelName: '@Value', defaultVal: null };
exports.complexValueInfo = [exports.complexValueType, exports.complexValueValue];
exports.styleSettingsTypes = [
    { propertyName: 'styleSettingsAppearance', modelName: 'AppearanceSettings', type: appearance_settings_1.AppearanceSettings },
    { propertyName: 'styleSettingsIcon', modelName: 'IconSettings', type: icon_settings_1.IconSettings },
    { propertyName: 'styleSettingsBar', modelName: 'StyleSettings', type: bar_style_settings_1.BarStyleSettings },
    { propertyName: 'styleSettingsRangeBar', modelName: 'BarStyleSettings', type: bar_style_settings_1.BarStyleSettings }
];
exports.styleSettings = { propertyName: 'styleSettings', displayName: 'DashboardWebStringId.StyleSettings' };
exports.barOptions = { propertyName: 'barOptions', modelName: 'BarOptions', displayName: 'DashboardWebStringId.ConditionalFormatting.BarOptions', type: format_condition_bar_options_1.FormatConditionBarOptions };
