// devexpress-dashboard/common/mobile-layout-extension/_dashboard-standalone-item.js
/**
* DevExpress Dashboard (_dashboard-standalone-item.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _interfaces_1 = require("../internal/_interfaces");
ko.components.register('dashboard-standalone-item', {
    viewModel: {
        createViewModel: function (params, componentInfo) {
            var element = componentInfo.element;
            return {
                dashboardItem: ko.unwrap(params.dashboardItem),
                dashboardContext: ko.unwrap(params.dashboardContext),
                localContext: ko.unwrap(params.localContext),
                sizeController: new _interfaces_1.SingleItemSizeController(element.parentElement, ko.unwrap(params.repaintRequest)),
                isStandalone: false
            };
        }
    },
    template: { element: 'dx-dashboard-standalone-item' }
});
