// devexpress-dashboard/designer/properties-controller/_properties-controller.js
/**
* DevExpress Dashboard (_properties-controller.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _accordion_tab_1 = require("./_accordion-tab");
var PropertiesController = (function () {
    function PropertiesController() {
        var _this = this;
        this._disposables = [];
        this.mainModel = ko.observable();
        this.secondaryModel = ko.observable();
        this.currentTab = ko.observable('');
        this.accordionDataSource = ko.observable([]);
        this.secondaryAccordionDataSource = ko.observable([]);
        this.computator = ko.computed(function () {
            var newTabs = [];
            var oldTabs = _this.accordionDataSource;
            if (_this.secondaryModel() && _this.secondaryModel().data) {
                newTabs = newTabs.concat(ko.unwrap(_this.secondaryModel().data.propertiesTabs));
                oldTabs = _this.secondaryAccordionDataSource;
            }
            else if (_this.mainModel() && _this.mainModel().data) {
                newTabs = newTabs.concat(ko.unwrap(_this.mainModel().data.propertiesTabs));
            }
            newTabs.forEach(function (tab, index) { if (!tab.orderNo)
                tab.orderNo = 100 + index; });
            newTabs = newTabs
                .sort(function (a, b) { return (a.orderNo) - (b.orderNo); });
            if ((newTabs.length === oldTabs().length)
                && newTabs.every(function (t) { return t instanceof _accordion_tab_1.AccordionTab; })
                && oldTabs().every(function (t) { return t instanceof _accordion_tab_1.AccordionTab; })) {
                var thesame = true;
                for (var i = 0; i < newTabs.length; i++) {
                    thesame = thesame && oldTabs()[i].name === newTabs[i].name;
                }
                if (thesame) {
                    for (var i = 0; i < newTabs.length; i++) {
                        oldTabs()[i].grabData(newTabs[i]);
                    }
                }
                else {
                    var tabsToDispose = oldTabs();
                    oldTabs(newTabs);
                    tabsToDispose.filter(function (tab) { return !newTabs.some(function (newTab) { return newTab === tab; }); }).forEach(function (tab) { return tab && tab.dispose(); });
                }
            }
            else {
                var tabsToDispose = oldTabs();
                oldTabs(newTabs);
                tabsToDispose.filter(function (tab) { return !newTabs.some(function (newTab) { return newTab === tab; }); }).forEach(function (tab) { return tab && tab.dispose(); });
            }
        });
        this.selectedIndex = ko.computed({
            read: function () {
                var newTabs = _this.accordionDataSource();
                var theSameTab = newTabs.filter(function (tab) { return (tab.category === _this.currentTab()) && tab.visible(); })[0];
                if (!theSameTab) {
                    theSameTab = newTabs.filter(function (tab) { return !tab.headerTemplate && tab.visible(); })[0];
                    theSameTab = theSameTab || newTabs[0];
                }
                return newTabs.indexOf(theSameTab);
            },
            write: function (index) {
                var newSelectedItem = _this.accordionDataSource()[index];
                if (!!newSelectedItem) {
                    _this.currentTab(newSelectedItem.category);
                }
                _this.secondaryModel(undefined);
            }
        }).extend({ notify: 'always', deferred: true });
        this.secondarySelectedIndex = ko.observable(0);
        this.processDataItemClick = function (data) {
            var model = data.item;
            if (!_this.mainModel() || !_this.mainModel().data || ko.unwrap(_this.mainModel().data.model) !== model) {
                data.click(model);
            }
            else {
                _this.mainModel(null);
            }
        };
        this._disposables.push(this.mainModel.subscribe(function () { return _this.secondaryModel(null); }, this, 'beforeChange'));
        this._disposables.push(this.mainModel.subscribe(function () {
            var newValue = _this.mainModel.peek();
            if (newValue && newValue.data && newValue.containingCollection) {
                _this._disposables.push(newValue.containingCollection.subscribe(function (changes) {
                    changes.forEach(function (arrayChange) {
                        var change = arrayChange;
                        if (change.status === 'deleted'
                            && _this.mainModel()
                            && ko.unwrap(_this.mainModel().data.model) === change.value) {
                            _this.mainModel(null);
                        }
                    });
                }, null, 'arrayChange'));
            }
            _this.secondaryModel(null);
        }));
        this._disposables.push(this.secondaryModel.subscribe(function () { return _this.secondarySelectedIndex(0); }, this, 'beforeChange'));
        this._disposables.push(this.secondaryModel.subscribe(function () {
            var newValue = _this.secondaryModel.peek();
            if (newValue && newValue.data && newValue.containingCollection) {
                _this._disposables.push(newValue.containingCollection.subscribe(function (changes) {
                    changes.forEach(function (arrayChange) {
                        var change = arrayChange;
                        if (change.status === 'deleted'
                            && _this.secondaryModel()
                            && ko.unwrap(_this.secondaryModel().data.model) === change.value) {
                            _this.secondaryModel(null);
                        }
                    });
                }, null, 'arrayChange'));
            }
        }));
        this._disposables.push(this.computator);
    }
    PropertiesController.prototype.dispose = function () {
        this.accordionDataSource().forEach(function (tab) { return tab && tab.dispose(); });
        this._disposables.forEach(function (d) {
            d.dispose();
        });
        this.selectedIndex.dispose();
    };
    return PropertiesController;
}());
exports.PropertiesController = PropertiesController;
