// devexpress-dashboard/viewer-parts/widgets/widgets-viewer/_css-class-names.js
/**
* DevExpress Dashboard (_css-class-names.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cssClassNames = {
    widgetsViewerScrollableHolder: 'dx-widgets-viewer-scrollable-holder',
    widgetViewerIdPrefix: 'dx-widgets-viewer-style_id',
    widgetViewerContainer: 'dx-widget-viewer-container',
    widgetViewerTable: 'widget-viewer-table',
    widgetViewerRow: 'widget-viewer-row',
    widgetViewerCell: 'widget-viewer-cell',
    cardItem: 'dx-cardItem',
    widgetItem: 'dx-widgetItem',
    cardTitle: 'dx-card-title',
    cardSubtitle: 'dx-card-subTitle',
    cardIndicator: 'dx-card-indicator',
    cardVariableValue1: 'dx-card-variableValue1',
    cardVariableValue2: 'dx-card-variableValue2',
    cardMainValue: 'dx-card-mainValue',
    cardSparkline: 'dx-card-sparkline',
    cardHiddenContainer: 'dx-card-item-hidden-container',
    cardNowrapHiddenContainer: 'dx-card-item-nowrap-hidden-container',
    layoutCardHiddenContainer: 'dx-layout-card-hidden-container',
    selectedItem: 'dx-selected-viewer-item',
    hoveredItem: 'dx-hovered-viewer-item'
};
exports.layoutCardCssClassNames = {
    cardScrollableHolder: 'dx-card-scrollable-holder',
    card: 'dx-flex-card-layout',
    content: 'dx-flex-card-layout-content',
    contentTopRows: 'dx-flex-card-layout-content-top-rows',
    contentBottomRows: 'dx-flex-card-layout-content-bottom-rows',
    row: 'dx-flex-card-layout-row',
    rowOnlyCenterElements: 'dx-flex-card-layout-row-center-only',
    rowLeftRightElements: 'dx-flex-card-layout-row-left-right',
    rowElement: 'dx-flex-card-layout-row-element',
    indicatorElement: 'dx-flex-card-layout-indicator-element',
    sparklineElement: 'dx-flex-card-layout-sparkline-row-element'
};
