// devexpress-dashboard/model/data-sources/metadata/_data-field.js
/**
* DevExpress Dashboard (_data-field.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _base_metadata_1 = require("../../metadata/_base-metadata");
exports.dataFieldDataMember = { propertyName: 'dataMember', modelName: 'DataMember' };
exports.dataFieldName = { propertyName: 'name', modelName: 'Name' };
exports.nodeType = { propertyName: 'nodeType', modelName: 'NodeType' };
exports.groupIndex = { propertyName: 'groupIndex', modelName: 'GroupIndex' };
exports.dataFieldChildNodes = { propertyName: 'childNodes', modelName: 'ChildNodes', array: true };
exports.dataFieldDisplayName = { propertyName: 'displayName', modelName: 'DisplayName' };
exports.dataFieldFieldType = {
    propertyName: 'fieldType', modelName: 'FieldType', defaultVal: 'Unknown', from: function (value) { return ko.observable(value === 'String' ? 'Text' : value); }
};
exports.dataFieldIsDataFieldNode = { propertyName: 'isDataFieldNode', modelName: 'IsDataFieldNode', from: _base_metadata_1.parseBool };
exports.dataFieldIsAggregate = { propertyName: 'isAggregate', modelName: 'IsAggregateCalculatedField', from: _base_metadata_1.parseBool };
exports.dataFieldIsDataMemberNode = { propertyName: 'isDataMemberNode', modelName: 'IsDataMemberNode', from: _base_metadata_1.parseBool };
exports.dataFieldIsConvertible = { propertyName: 'isConvertible', modelName: 'IsConvertible', from: _base_metadata_1.parseBool };
exports.dataFieldIsComparable = { propertyName: 'isComparable', modelName: 'IsComparable', from: _base_metadata_1.parseBool };
exports.dataFieldIsList = { propertyName: 'isList', modelName: 'IsList', from: _base_metadata_1.parseBool };
exports.dataFieldIsCorruptedCalField = { propertyName: 'isCorruptedCalcField', modelName: 'IsCorruptedCalculatedField', from: _base_metadata_1.parseBool };
exports.dataFieldSerializationsInfo = [exports.dataFieldDataMember, exports.dataFieldName, exports.dataFieldChildNodes, exports.dataFieldDisplayName, exports.dataFieldFieldType, exports.dataFieldIsDataMemberNode, exports.dataFieldIsDataFieldNode, exports.dataFieldIsConvertible, exports.dataFieldIsComparable, exports.nodeType, exports.groupIndex, exports.dataFieldIsAggregate, exports.dataFieldIsList, exports.dataFieldIsCorruptedCalField];
