// devexpress-dashboard/common/extensions/viewer-api-extension.js
/**
* DevExpress Dashboard (viewer-api-extension.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _drill_through_data_wrapper_1 = require("../../data/drill-through-data/_drill-through-data-wrapper");
var _default_1 = require("../../data/localization/_default");
var _common_1 = require("../../data/_common");
var model_1 = require("../../model");
var disposable_object_1 = require("../../model/disposable-object");
var _array_utils_1 = require("../../model/internal/_array-utils");
var _helper_classes_1 = require("../../model/internal/_helper-classes");
var _knockout_utils_1 = require("../../model/internal/_knockout-utils");
var data_dashboard_item_1 = require("../../model/items/data-dashboard-item");
var date_filter_item_1 = require("../../model/items/filter-items/date-filter-item");
var range_filter_item_1 = require("../../model/items/range-filter/range-filter-item");
var _events_helper_1 = require("../../viewer-parts/viewer/_events-helper");
var control_options_1 = require("../control-options");
var _underlying_data_provider_1 = require("../data/_underlying-data-provider");
var utils_1 = require("../internal/utils");
var _options_manager_1 = require("../internal/_options-manager");
var notificator_1 = require("../notification-controller/notificator");
var name = 'viewer-api';
var ViewerApiExtension = (function (_super) {
    __extends(ViewerApiExtension, _super);
    function ViewerApiExtension(dashboardControl, options) {
        if (options === void 0) { options = {}; }
        var _this = _super.call(this) || this;
        _this.dashboardControl = dashboardControl;
        _this.name = name;
        _this._viewerItems = {};
        _this._optionManager = new _options_manager_1.OptionsManager();
        _this._defaultOptions = {
            onItemClick: function (args) { },
            onItemSelectionChanged: function (args) { },
            onItemWidgetCreated: function (args) { },
            onItemWidgetUpdating: function (args) { },
            onItemWidgetUpdated: function (args) { },
            onItemElementCustomColor: function (args) { },
            onItemVisualInteractivity: function (args) { },
            onItemMasterFilterStateChanged: function (args) { },
            onItemDrillDownStateChanged: function (args) { },
            onItemActionAvailabilityChanged: function (args) { },
            onItemCaptionToolbarUpdated: function (args) { },
            onDashboardTitleToolbarUpdated: function (args) { },
            onSelectedTabPageChanged: function (args) { },
            onDynamicLookUpValuesLoaded: function (args) { },
            onItemBeginUpdate: function (args) { },
            onItemEndUpdate: function (args) { },
            onDashboardBeginUpdate: function (args) { },
            onDashboardEndUpdate: function (args) { },
            onItemWidgetOptionsPrepared: function (args) { }
        };
        _this._dashboardDisposables = [];
        _this._title = ko.observable(null);
        _this._viewerItemCreated = function (item, viewerItem) {
            if (!!viewerItem) {
                if (item instanceof data_dashboard_item_1.DataDashboardItem) {
                    viewerItem.itemClick.add(_this._raiseItemClick);
                    viewerItem.itemSelectionChanged.add(_this._raiseItemSelectionChanged);
                    viewerItem.clearMasterFilter.add(_this._raiseClearMasterFilter);
                }
                if (!!viewerItem['itemElementCustomColor']) {
                    viewerItem['itemElementCustomColor'].add(_this._raiseItemElementCustomColor);
                }
                viewerItem.itemWidgetCreated.add(_this._raiseItemWidgetCreated);
                viewerItem.itemWidgetUpdating.add(_this._raiseItemWidgetUpdating);
                viewerItem.itemWidgetUpdated.add(_this._raiseItemWidgetUpdated);
                viewerItem.itemWidgetOptionsPrepared.add(_this._raiseItemWidgetOptionsPrepared);
                viewerItem.itemCaptionToolbarUpdated.add(_this._raiseItemCaptionToolbarUpdated);
            }
            if (!_this._viewerItems[item.componentName()]) {
                _this._viewerItems[item.componentName()] = [];
            }
            if (_this._viewerItems[item.componentName()].indexOf(viewerItem) === -1) {
                _this._viewerItems[item.componentName()].push(viewerItem);
            }
        };
        _this._viewerItemDispose = function (item, viewerItem) {
            if (!!viewerItem) {
                if (item instanceof data_dashboard_item_1.DataDashboardItem) {
                    viewerItem.itemClick.remove(_this._raiseItemClick);
                    viewerItem.itemSelectionChanged.remove(_this._raiseItemSelectionChanged);
                    viewerItem.clearMasterFilter.remove(_this._raiseClearMasterFilter);
                }
                if (!!viewerItem['itemElementCustomColor']) {
                    viewerItem['itemElementCustomColor'].remove(_this._raiseItemElementCustomColor);
                }
                viewerItem.itemWidgetCreated.remove(_this._raiseItemWidgetCreated);
                viewerItem.itemWidgetUpdating.remove(_this._raiseItemWidgetUpdating);
                viewerItem.itemWidgetUpdated.remove(_this._raiseItemWidgetUpdated);
                viewerItem.itemCaptionToolbarUpdated.remove(_this._raiseItemCaptionToolbarUpdated);
            }
            if (_this._viewerItems[item.componentName()]) {
                var index = _this._viewerItems[item.componentName()].indexOf(viewerItem);
                if (index > -1) {
                    _this._viewerItems[item.componentName()].splice(index, 1);
                }
            }
        };
        _this._beforeApplyViewerItemOptions = function (item, options, isCreation, customInteractivityOptions) {
            if (item instanceof data_dashboard_item_1.DataDashboardItem) {
                _this._optionManager.events.raise('itemVisualInteractivity', _events_helper_1.createItemInteractivityEventArgs(item, customInteractivityOptions));
            }
        };
        _this._raiseItemActionAvailabilityChanged = function (item) {
            _this._optionManager.events.raise('itemActionAvailabilityChanged', {
                itemName: item.componentName(),
                dashboardItem: item,
            });
        };
        _this._raiseItemClick = function (itemName, dataPoint) {
            _this._optionManager.events.raise('itemClick', _events_helper_1.createItemClickEventArgs(_this._getDataItem(itemName), dataPoint, _this.requestUnderlyingData));
        };
        _this._raiseItemSelectionChanged = function (itemName, tuples) {
            _this._optionManager.events.raise('itemSelectionChanged', _events_helper_1.createItemSelectionChangedEventArgs(_this._getDataItem(itemName), tuples));
        };
        _this._raiseItemWidgetCreated = function (name, viewControl) {
            var item = _this._getItem(name);
            _this._optionManager.events.raise('itemWidgetCreated', _events_helper_1.createWidgetEventArgs(item, viewControl));
        };
        _this._raiseItemWidgetUpdating = function (name, viewControl) {
            var item = _this._getItem(name);
            _this._optionManager.events.raise('itemWidgetUpdating', _events_helper_1.createWidgetEventArgs(item, viewControl));
        };
        _this._raiseItemWidgetUpdated = function (name, viewControl) {
            var item = _this._getItem(name);
            _this._optionManager.events.raise('itemWidgetUpdated', _events_helper_1.createWidgetEventArgs(item, viewControl));
        };
        _this._raiseItemWidgetOptionsPrepared = function (name, options) {
            var item = _this._getItem(name);
            _this._optionManager.events.raise('itemWidgetOptionsPrepared', _events_helper_1.createWidgetOptionsEventArgs(item, options));
        };
        _this._raiseItemCaptionToolbarUpdated = function (name, options) {
            var item = _this._getItem(name);
            _this._optionManager.events.raise('itemCaptionToolbarUpdated', { itemName: name, dashboardItem: item, options: options });
        };
        _this._raiseTitleToolbarUpdated = function (options) {
            _this._optionManager.events.raise('dashboardTitleToolbarUpdated', { dashboard: _this.dashboardControl.dashboard(), options: options });
        };
        _this._raiseItemElementCustomColor = function (itemName, eventArgs) {
            _this._optionManager.events.raise('itemElementCustomColor', _events_helper_1.createItemElementCustomColorEventArgs(_this._getDataItem(itemName), eventArgs));
        };
        _this._raiseItemVisualInteractivity = function (itemName, interactivityOptions) {
            _this._optionManager.events.raise('itemVisualInteractivity', _events_helper_1.createItemInteractivityEventArgs(_this._getDataItem(itemName), interactivityOptions));
        };
        _this._raiseClearMasterFilter = function (itemName) {
            var item = _this._getItem(itemName);
            _this._optionManager.events.raise('itemMasterFilterStateChanged', {
                itemName: item.componentName(),
                dashboardItem: item,
                values: null
            });
        };
        _this.requestUnderlyingData = function (itemName, args, onCompleted) {
            var dataDashboardItem = _this._getDataItem(itemName);
            var raiseOnCompleted = function (underlyingData) {
                var drillThroughData = new _drill_through_data_wrapper_1.DrillThroughDataWrapper(underlyingData);
                drillThroughData.initialize();
                onCompleted(drillThroughData);
            };
            var provider = new _underlying_data_provider_1.UnderlyingDataProvider(_this.dashboardControl._serviceClient());
            provider.requestUnderlyingData(dataDashboardItem, args)
                .done(raiseOnCompleted)
                .fail(function (request) {
                var errorMessage = notificator_1.NotificationController._getErrorTextFromResponse(request);
                if (!errorMessage) {
                    errorMessage = _default_1.getLocalizationById('DashboardWebStringId.Notification.AttemptToGetUnderlyingData');
                }
                raiseOnCompleted({ ErrorMessage: errorMessage });
            });
        };
        _this._optionManager.initialize(_this._defaultOptions, options, _this);
        _this.toDispose.apply(_this, _knockout_utils_1.subscribeWithPrev(_this._title, function (oldValue, newValue) {
            if (oldValue) {
                oldValue.onUpdated.remove(_this._raiseTitleToolbarUpdated);
            }
            if (newValue) {
                newValue.onUpdated.add(_this._raiseTitleToolbarUpdated);
            }
        }));
        _this.toDispose(ko.computed(function () {
            var newDashboard = dashboardControl.dashboard();
            _this._dashboardDisposables.forEach(function (d) { return d.dispose(); });
            _this._dashboardDisposables = [];
            if (newDashboard) {
                newDashboard
                    .items()
                    .filter(function (item) { return item instanceof model_1.TabContainerItem; })
                    .forEach(function (tabContainer) {
                    tabContainer._activePageChanged = function (prevPageName, pageName) {
                        _this._raiseSelectedTabPageChanged(tabContainer.componentName(), prevPageName, pageName);
                    };
                });
                newDashboard._dataDashboardItems().forEach(function (item) {
                    var itemName = item.componentName.peek();
                    _this._raiseItemActionAvailabilityChanged(item);
                    var disposables = _knockout_utils_1.subscribeWithPrev(item._actions, function (prevActions, actions) {
                        if (!_array_utils_1.arrayEquals(prevActions, actions)) {
                            _this._raiseItemActionAvailabilityChanged(item);
                        }
                    });
                    disposables.push(_knockout_utils_1.subscribeArrayChange(item._drillDownValues, {
                        added: function (v) {
                            _this._optionManager.events.raise('itemDrillDownStateChanged', {
                                itemName: itemName,
                                dashboardItem: item,
                                action: 'Down',
                                values: item._drillDownValues()
                            });
                        },
                        deleted: function (v) {
                            _this._optionManager.events.raise('itemDrillDownStateChanged', {
                                itemName: itemName,
                                dashboardItem: item,
                                action: 'Up',
                                values: item._drillDownValues()
                            });
                        }
                    }));
                    disposables.push(item._actualSelectionValues.subscribe(function (newValue) {
                        _this._optionManager.events.raise('itemMasterFilterStateChanged', {
                            itemName: itemName,
                            dashboardItem: item,
                            values: newValue
                        });
                    }));
                    Array.prototype.push.apply(_this._dashboardDisposables, disposables);
                });
            }
        }));
        return _this;
    }
    ViewerApiExtension.prototype._checkIsRangeFilterItem = function (itemName) {
        var item = this._getDataItem(itemName);
        if (!(item instanceof range_filter_item_1.RangeFilterItem) && !(item instanceof date_filter_item_1.DateFilterItem)) {
            throw new Error('Action is called for an unsupported dashboard item. This action can be performed only for Range Filter and Date Filter.');
        }
    };
    ViewerApiExtension.prototype._raiseSelectedTabPageChanged = function (tabContainerName, prevPageName, pageName) {
        this._optionManager.events.raise('selectedTabPageChanged', {
            tabContainerName: tabContainerName,
            selectedPage: pageName,
            previousPage: prevPageName
        });
    };
    ViewerApiExtension.prototype.start = function () {
        this.dashboardControl._dashboardContext.viewerItemCreated.add(this._viewerItemCreated);
        this.dashboardControl._dashboardContext.viewerItemDispose.add(this._viewerItemDispose);
        this.dashboardControl._dashboardContext.beforeApplyViewerItemOptions.add(this._beforeApplyViewerItemOptions);
    };
    ViewerApiExtension.prototype.stop = function () {
        this.dashboardControl._dashboardContext.viewerItemCreated.remove(this._viewerItemCreated);
        this.dashboardControl._dashboardContext.viewerItemDispose.remove(this._viewerItemDispose);
        this.dashboardControl._dashboardContext.beforeApplyViewerItemOptions.remove(this._beforeApplyViewerItemOptions);
    };
    ViewerApiExtension.prototype._getItemCore = function (itemId, findItem) {
        var dashboard = this.dashboardControl.dashboard();
        if (dashboard) {
            var item = findItem(dashboard);
            if (item) {
                return item;
            }
            else {
                throw Error("The item with the '" + itemId + "' name does not exist");
            }
        }
        else {
            throw new Error('Cannot perform operation because the dashboard is not loaded');
        }
    };
    ViewerApiExtension.prototype._getItem = function (itemId) {
        return this._getItemCore(itemId, function (dashboard) { return dashboard._allItems().filter(function (item) { return item.componentName() == itemId; })[0]; });
    };
    ViewerApiExtension.prototype._getDataItem = function (itemName) {
        return this._getItemCore(itemName, function (dashboard) { return dashboard._findDataItem(itemName); });
    };
    ViewerApiExtension.prototype.getCurrentRange = function (itemName) {
        this._checkIsRangeFilterItem(itemName);
        var item = this._getDataItem(itemName);
        var selection = item._actualSelectionValues() && item._actualSelectionValues()[0] || item._getEntireRange();
        return selection && selection.length > 0 ? {
            minimum: selection[0],
            maximum: selection[1]
        } : null;
    };
    ViewerApiExtension.prototype.getEntireRange = function (itemName) {
        this._checkIsRangeFilterItem(itemName);
        var entireRange = this._getDataItem(itemName)._getEntireRange();
        return entireRange.length > 0 ? {
            minimum: entireRange[0],
            maximum: entireRange[1]
        } : null;
    };
    ViewerApiExtension.prototype.setRange = function (itemName, range) {
        this._checkIsRangeFilterItem(itemName);
        this.setMasterFilter(itemName, [[range.minimum, range.maximum]]);
    };
    ViewerApiExtension.prototype.setPredefinedRange = function (itemName, dateTimePeriodName) {
        this._checkIsRangeFilterItem(itemName);
        var rangeFilter = this._getDataItem(itemName);
        var periods = rangeFilter.dateTimePeriods().filter(function (period) { return period.name() === dateTimePeriodName; });
        if (periods.length === 0) {
            throw new Error("The predefined range with the '" + dateTimePeriodName + "' name does not exist");
        }
        rangeFilter._processItemSetPredefinedPeriod(dateTimePeriodName);
    };
    ViewerApiExtension.prototype.getAvailablePredefinedRanges = function (itemName) {
        this._checkIsRangeFilterItem(itemName);
        return this._getDataItem(itemName).dateTimePeriods().map(function (period) { return period.name(); });
    };
    ViewerApiExtension.prototype.getCurrentPredefinedRange = function (itemName) {
        this._checkIsRangeFilterItem(itemName);
        var period = this._getDataItem(itemName).currentSelectedDateTimePeriodName();
        return period ? period : '';
    };
    ViewerApiExtension.prototype.getCurrentSelection = function (itemName) {
        _helper_classes_1.Guard.isNotNull(itemName, 'itemName');
        var itemData = this._getDataItem(itemName)._getItemData(), tuples = [], viewerItem = this._getViewerItem(itemName);
        var selectedTuples = viewerItem && viewerItem.getSelectedTuples() || null;
        if (selectedTuples) {
            tuples = selectedTuples.map(function (selectedTuple) { return itemData.createTuple(selectedTuple); });
        }
        return tuples;
    };
    ViewerApiExtension.prototype.canSetMasterFilter = function (itemName) {
        return this._getDataItem(itemName)._actions().indexOf(_common_1.viewerActions.setMasterFilter) !== -1;
    };
    ViewerApiExtension.prototype.canClearMasterFilter = function (itemName) {
        return this._getDataItem(itemName)._actions().indexOf(_common_1.viewerActions.clearMasterFilter) !== -1;
    };
    ViewerApiExtension.prototype.canPerformDrillDown = function (itemName) {
        return this._getDataItem(itemName)._actions().indexOf(_common_1.viewerActions.drillDown) !== -1;
    };
    ViewerApiExtension.prototype.canPerformDrillUp = function (itemName) {
        return this._getDataItem(itemName)._actions().indexOf(_common_1.viewerActions.drillUp) !== -1;
    };
    ViewerApiExtension.prototype.getItemData = function (itemName) {
        return this._getDataItem(itemName)._getItemData();
    };
    ViewerApiExtension.prototype.getCurrentFilterValues = function (itemName) {
        return this._getDataItem(itemName)._getCurrentFilterValues();
    };
    ViewerApiExtension.prototype.getAvailableFilterValues = function (itemName) {
        return this._getDataItem(itemName)._getAvailableFilterValues(itemName);
    };
    ViewerApiExtension.prototype.getCurrentDrillDownValues = function (itemName) {
        return this._getDataItem(itemName)._getCurrentDrillDownValues();
    };
    ViewerApiExtension.prototype.getAvailableDrillDownValues = function (itemName) {
        return this._getDataItem(itemName)._getAvailableDrillDownValues(itemName);
    };
    ViewerApiExtension.prototype.setMasterFilter = function (itemName, values) {
        this._getDataItem(itemName)._performSetMasterFilter(values);
    };
    ViewerApiExtension.prototype.clearMasterFilter = function (itemName) {
        this._getDataItem(itemName)._performClearMasterFilter();
    };
    ViewerApiExtension.prototype.performDrillDown = function (itemName, value) {
        this._getDataItem(itemName)._performDrillDown(value);
    };
    ViewerApiExtension.prototype.performDrillUp = function (itemName) {
        this._getDataItem(itemName)._performDrillUp();
    };
    ViewerApiExtension.prototype.getAvailableActions = function (itemName) {
        return this._getDataItem(itemName)._getAvailableActions();
    };
    ViewerApiExtension.prototype.updateItemCaptionToolbar = function (itemName) {
        var _this = this;
        var dashboardItemNames = itemName ?
            this.dashboardControl.dashboard().findItem(itemName) && [itemName] || []
            :
                this.dashboardControl
                    .dashboard()
                    .items()
                    .concat(this.dashboardControl.dashboard().groups())
                    .map(function (item) { return item.componentName(); });
        if (itemName && dashboardItemNames.length === 0) {
            throw new Error("The item with the '" + itemName + "' name does not exist");
        }
        dashboardItemNames.forEach(function (itemName) {
            _this._viewerItems[itemName] && _this._viewerItems[itemName].forEach(function (viewerItem) { return viewerItem.updateCaptionToolbar(); });
        });
    };
    ViewerApiExtension.prototype.updateDashboardTitleToolbar = function () {
        if (this._title()) {
            this._title().update();
        }
    };
    ViewerApiExtension.prototype.setSelectedTabPage = function (tabPageName) {
        var dashboard = this.dashboardControl.dashboard();
        if (dashboard) {
            var tabPage = utils_1.findItemForApi(this.dashboardControl.dashboard(), tabPageName, model_1.DashboardTabPage);
            var tabContainer = this._findParentTabContainer(tabPageName);
            if (tabContainer && tabPage)
                tabContainer._activeTabPage(tabPage);
        }
    };
    ViewerApiExtension.prototype.setSelectedTabPageIndex = function (tabContainerName, index) {
        var dashboard = this.dashboardControl.dashboard();
        if (dashboard) {
            var tabContainer = utils_1.findItemForApi(this.dashboardControl.dashboard(), tabContainerName, model_1.TabContainerItem);
            if (index < 0 || index > tabContainer.tabPages().length - 1) {
                throw Error("The '" + tabContainerName + "' tab container does not contain a tab page with the following index: " + index);
            }
            var tabContainerLayoutItem = this.dashboardControl.dashboard().layout().findLayoutItem(tabContainer);
            if (tabContainerLayoutItem) {
                this.setSelectedTabPage(tabContainerLayoutItem.childNodes()[index].dashboardItem());
            }
        }
    };
    ViewerApiExtension.prototype.getSelectedTabPageIndex = function (tabContainerName) {
        var dashboard = this.dashboardControl.dashboard();
        if (dashboard) {
            var tabContainer = utils_1.findItemForApi(this.dashboardControl.dashboard(), tabContainerName, model_1.TabContainerItem);
            var tabContainerLayoutItem = this.dashboardControl.dashboard().layout().findLayoutItem(tabContainer);
            var activeTabPageLayoutItem = this.dashboardControl.dashboard().layout().findLayoutItem(tabContainer._activeTabPage());
            if (tabContainerLayoutItem && tabContainerLayoutItem) {
                return tabContainerLayoutItem.childNodes().indexOf(activeTabPageLayoutItem);
            }
        }
        return -1;
    };
    ViewerApiExtension.prototype.getSelectedTabPage = function (tabContainerName) {
        var dashboard = this.dashboardControl.dashboard();
        if (dashboard) {
            var tabContainer = utils_1.findItemForApi(this.dashboardControl.dashboard(), tabContainerName, model_1.TabContainerItem);
            return tabContainer._activeTabPage() ? tabContainer._activeTabPage().componentName() : '';
        }
        return '';
    };
    ViewerApiExtension.prototype._findParentTabContainer = function (tabPageName) {
        var tabContainers = this.dashboardControl.dashboard().items().filter(function (item) { return item instanceof model_1.TabContainerItem; });
        var parentContainer;
        tabContainers.forEach(function (container) {
            var pages = container.tabPages().filter(function (page) { return page.componentName() === tabPageName; });
            if (pages.length > 0) {
                parentContainer = container;
            }
        });
        if (!parentContainer) {
            throw new Error("The tab container item with the page'" + tabPageName + "' name does not exist");
        }
        return parentContainer;
    };
    ViewerApiExtension.prototype._getViewerItem = function (itemName) {
        var viewerItems = this._viewerItems[itemName] ? this._viewerItems[itemName].filter(function (viewer) { return viewer.hasWidget; }) : [];
        return viewerItems.length > 0 ? viewerItems[0] : undefined;
    };
    return ViewerApiExtension;
}(disposable_object_1.DisposableObject));
exports.ViewerApiExtension = ViewerApiExtension;
control_options_1.defaultExtensions[name] = function (dashboardControl, options) { return new ViewerApiExtension(dashboardControl, options); };
