// devexpress-dashboard/viewer-parts/viewer-items/range-selector-item/_range-selector-item.js
/**
* DevExpress Dashboard (_range-selector-item.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var range_selector_1 = require("devextreme/viz/range_selector");
var $ = require("jquery");
var special_values_1 = require("../../../data/special-values");
var _chart_helper_1 = require("../../../data/_chart-helper");
var _common_1 = require("../../../data/_common");
var _localization_ids_1 = require("../../../data/_localization-ids");
var _localizer_1 = require("../../../data/_localizer");
var _utils_1 = require("../../../data/_utils");
var caption_toolbar_options_1 = require("../../widgets/caption-toolbar/caption-toolbar-options");
var _caption_toolbar_css_classes_1 = require("../../widgets/caption-toolbar/_caption-toolbar-css-classes");
var _custom_time_period_dialog_1 = require("../../widgets/dialogs/_custom-time-period-dialog");
var _render_helper_1 = require("../../widgets/_render-helper");
var _base_item_1 = require("../_base-item");
var _range_filter_selection_validator_1 = require("./_range-filter-selection-validator");
var DayIntervalWidthInPixels = 25, MonthIntervalWidthInPixels = 40, QuarterIntervalWidthInPixels = 20;
var rangeSelectorItem = (function (_super) {
    __extends(rangeSelectorItem, _super);
    function rangeSelectorItem(container, options) {
        var _this = _super.call(this, container, options) || this;
        _this.itemElementCustomColor = $.Callbacks();
        _this.timePeriodMenuSelectedIndex = undefined;
        return _this;
    }
    Object.defineProperty(rangeSelectorItem.prototype, "_isBottomFloatingToolbarPosition", {
        get: function () {
            return true;
        },
        enumerable: true,
        configurable: true
    });
    rangeSelectorItem.prototype._getCustomTimePeriodDialog = function () {
        var _this = this;
        if (!this._customTimePeriodDialog) {
            this._customTimePeriodDialog = new _custom_time_period_dialog_1.customTimePeriodDialog({
                container: $(this.container).closest('.' + _caption_toolbar_css_classes_1.cssClasses.dashboardContainer).get(0),
                setRange: function (range) {
                    _this.selected.fire(_this.getName(), _common_1.viewerActions.setMasterFilter, [[range.startValue, range.endValue]]);
                }
            });
        }
        return this._customTimePeriodDialog;
    };
    rangeSelectorItem.prototype._initializeData = function (newOptions) {
        _super.prototype._initializeData.call(this, newOptions);
        if (this.dataController) {
            this.dataController.elementCustomColor = $.proxy(this._elementCustomColor, this);
        }
    };
    rangeSelectorItem.prototype._clearSelectionUnsafe = function () {
        this._setRange([]);
    };
    rangeSelectorItem.prototype._clearSelectionBase = function () {
        this.timePeriodMenuSelectedIndex = undefined;
    };
    rangeSelectorItem.prototype._setSelectionUnsafe = function (values) {
        this._setRange(values && values.length > 0 ? values[0] : []);
    };
    rangeSelectorItem.prototype._getCurrentRange = function () {
        var value = this._getSelectedValues();
        var startValue, endValue;
        if (value && value !== null) {
            startValue = value[0];
            endValue = value[1];
        }
        if (!_utils_1.type.isDefined(startValue) || !_utils_1.type.isDefined(endValue)) {
            var entireRange = this._getEntireRange();
            if (!_utils_1.type.isDefined(startValue))
                startValue = entireRange.startValue;
            if (!_utils_1.type.isDefined(endValue))
                endValue = entireRange.endValue;
        }
        return new rangeFilterSelection({ startValue: startValue, endValue: endValue });
    };
    rangeSelectorItem.prototype.getEntireRange = function () {
        return new rangeFilterSelection(this._getEntireRange());
    };
    rangeSelectorItem.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        var customPeriodForm = this._getCustomTimePeriodDialog();
        if (customPeriodForm) {
            customPeriodForm.dispose();
        }
        this.rangeSelectorViewer && this.rangeSelectorViewer.dispose();
    };
    rangeSelectorItem.prototype._setRange = function (range) {
        this._setSelectedValues([range]);
        if (this.hasWidget) {
            this._lock();
            this.rangeSelectorViewer.setValue(range);
            this._unlock();
        }
        else {
            this.selected.fire(this.getName(), _common_1.viewerActions.setMasterFilter, [range]);
        }
    };
    rangeSelectorItem.prototype._setPredefinedRange = function (dateTimePeriodName) {
        var dateTimePeriods = this.options.ViewModel.DateTimePeriods;
        var period = dateTimePeriods.filter(function (dateTimePeriod) { return dateTimePeriod.Name === dateTimePeriodName; })[0];
        this.timePeriodMenuSelectedIndex = !!period ? dateTimePeriods.indexOf(period) : undefined;
        this.updateCaptionToolbar();
    };
    rangeSelectorItem.prototype._getAvailablePredefinedRanges = function () {
        return this.options.ViewModel.DateTimePeriods.map(function (period) { return period.Name; });
    };
    rangeSelectorItem.prototype.renderContentUnsafe = function (element, changeExisting, afterRenderCallback) {
        var that = this, options = this._getRangeSelectorViewerOptions();
        var isFirstInitialization = !changeExisting || !that.rangeSelectorViewer;
        this._lock();
        this._raiseItemWidgetOptionsPrepared(options);
        try {
            if (isFirstInitialization) {
                this.rangeSelectorViewer = new range_selector_1.default(element, options);
            }
            else {
                this.rangeSelectorViewer.option(options);
            }
        }
        finally {
            this._unlock();
        }
        this.timePeriodMenuSelectedIndex = this.options.ViewModel.SelectedPeriodIndex;
        return false;
    };
    rangeSelectorItem.prototype._isBorderRequired = function () {
        return false;
    };
    rangeSelectorItem.prototype._getContainerPositionUnsafe = function () {
        var position = _super.prototype._getContainerPositionUnsafe.call(this), buttonOffset = this._getButtonOffset(true);
        position.offsetY = position.height - buttonOffset.top;
        return position;
    };
    rangeSelectorItem.prototype._getFirstDayOfWeek = function () {
        var rangeDataController = this.dataController;
        var argumentFormat = rangeDataController.getArgumentFormat();
        return argumentFormat ? argumentFormat.firstDayOfWeek : null;
    };
    rangeSelectorItem.prototype._getRangeSelectorViewerOptions = function () {
        var that = this;
        var rangeDataController = that.dataController;
        var viewModel = that.options.ViewModel, argument = viewModel.Argument, scaleType = rangeSelectorItem._getScaleType(argument), seriesTemplates = viewModel.SeriesTemplates, selectedValues = this._getSelectedValues(), dataSourceAndSeries = rangeDataController.getDataSourceAndSeries(that._isEncodeHtml()), tickMarkInterval = undefined, animation = that._getAnimationOptions(), options = {
            scale: {
                axisDivisionFactor: null,
                type: null,
                marker: {
                    visible: false
                }
            },
            encodeHtml: that._isEncodeHtml(),
            margin: {
                top: 10,
                bottom: 15
            }
        };
        if (scaleType !== 'discrete') {
            var entireRange = this._getEntireRangeByDataSource(dataSourceAndSeries.dataSource);
            options.scale.startValue = entireRange.startValue;
            options.scale.endValue = entireRange.endValue;
        }
        else {
            options.scale.startValue = null;
            options.scale.endValue = null;
        }
        if (!this._isValidValues(selectedValues)) {
            options.value = [selectedValues[0], selectedValues[1]];
        }
        var argumentFormat = rangeDataController.getArgumentFormat();
        options.scale.label = {
            format: argumentFormat
        };
        options.scale.valueType = argument.Type == 'DateTime' ? 'datetime' : 'numeric';
        options.sliderMarker = {
            format: argumentFormat
        };
        var dataSource = dataSourceAndSeries.dataSource;
        options.scale.type = scaleType;
        if (argument) {
            if ((argument.DateTimeGroupInterval === 'Year' && argument.Type === 'DateTime') ||
                !argument.IsContinuousDateTimeScale) {
                tickMarkInterval = rangeDataController.isDiscreteArgument() || argument.Type == 'Integer' ? 1 : _chart_helper_1.chartHelper.convertPresentationUnit(argument);
                if (tickMarkInterval) {
                    var firstDayOfWeek = this._getFirstDayOfWeek();
                    if (tickMarkInterval === 'week') {
                        if (Number.isInteger(firstDayOfWeek)) {
                            options.scale.workWeek = [firstDayOfWeek];
                        }
                        else {
                            throw new Error('First Day Of Week is not defined for range filter agrument');
                        }
                    }
                    options.scale.minorTick = {
                        visible: false
                    };
                    options.behavior = {
                        snapToTicks: true,
                        animationEnabled: animation.enabled
                    };
                    options.scale.minRange = tickMarkInterval;
                    if (scaleType === 'semidiscrete') {
                        options.scale.axisDivisionFactor = {
                            day: DayIntervalWidthInPixels,
                            month: MonthIntervalWidthInPixels,
                            quarter: QuarterIntervalWidthInPixels
                        };
                    }
                }
            }
            else {
                options.behavior = {
                    snapToTicks: false,
                    animationEnabled: animation.enabled
                };
            }
        }
        options.dataSource = dataSource;
        var chartSeries = dataSourceAndSeries.series;
        if (scaleType === 'discrete') {
            chartSeries.forEach(function (series) {
                series.ignoreEmptyPoints = true;
            });
        }
        options.chart = {
            series: chartSeries,
            commonSeriesSettings: {
                type: seriesTemplates && seriesTemplates.length > 0 ? (_chart_helper_1.chartHelper.convertSeriesType(seriesTemplates[0].SeriesType)) : null
            },
            palette: _render_helper_1.RenderHelper.getDefaultPalette()
        };
        options.onValueChanged = that._getSelectedRangeChangedHandler();
        return options;
    };
    rangeSelectorItem.prototype._getEntireRange = function () {
        var rangeDataController = this.dataController;
        var dataSourceAndSeries = rangeDataController.getDataSourceAndSeries(this._isEncodeHtml());
        return this._getEntireRangeByDataSource(dataSourceAndSeries.dataSource);
    };
    rangeSelectorItem.prototype._getEntireRangeByDataSource = function (dataSource) {
        var _this = this;
        var range;
        var lastDataItemIndex = dataSource ? _utils_1.findLastIndex(dataSource, function (item) { return _this._checkAxisXValue(item.x); }) : -1;
        if (lastDataItemIndex > 0) {
            var firstDataItemIndex = _utils_1.findIndex(dataSource, function (item) { return _this._checkAxisXValue(item.x); });
            var isQualitativeArgument = this.dataController.isQualitativeArgument();
            range = {
                startValue: isQualitativeArgument ? firstDataItemIndex : dataSource[firstDataItemIndex].x,
                endValue: isQualitativeArgument ? lastDataItemIndex : dataSource[lastDataItemIndex].x
            };
        }
        else {
            range = {
                startValue: null,
                endValue: null
            };
        }
        return range;
    };
    rangeSelectorItem.prototype._checkAxisXValue = function (value) {
        return _utils_1.type.isDefined(value) && Object.keys(special_values_1.specialValues).filter(function (specialValueName) { return special_values_1.specialValues[specialValueName] === value; }).length === 0;
    };
    rangeSelectorItem.prototype._getSelectedValues = function () {
        var allSelectedValues = _super.prototype._getSelectedValues.call(this);
        return allSelectedValues ? allSelectedValues[0] : null;
    };
    rangeSelectorItem.prototype._isValidValues = function (values) {
        if (values && values.length) {
            var startValue = values[0], endValue = values[1];
            if (_range_filter_selection_validator_1.RangeFilterSelectionValidator.isValidValue(startValue) && _range_filter_selection_validator_1.RangeFilterSelectionValidator.isValidValue(endValue))
                return false;
        }
        return true;
    };
    rangeSelectorItem.prototype._getSliderMarkerFormat = function () {
        return this.dataController.isSingleArgument() ? this.dataController.getSingleArgumentDimensionFormat() : undefined;
    };
    rangeSelectorItem._getScaleType = function (argument) {
        if (argument) {
            var dateTimeGroupInterval = argument.DateTimeGroupInterval;
            switch (argument.Type) {
                case 'String':
                    var groupIntervalScaleType = rangeSelectorItem._getStringScaleTypeByGroupInterval(dateTimeGroupInterval);
                    return _utils_1.type.isDefined(groupIntervalScaleType) ? groupIntervalScaleType : 'discrete';
                case 'Integer':
                    return 'semidiscrete';
                case 'DateTime':
                    return rangeSelectorItem._getDateTimeScaleType(dateTimeGroupInterval);
                default:
                    return null;
            }
        }
    };
    rangeSelectorItem._getStringScaleTypeByGroupInterval = function (dateTimeGroupInterval) {
        switch (dateTimeGroupInterval) {
            case 'Year':
            case 'Quarter':
            case 'Month':
            case 'Day':
            case 'Hour':
            case 'Minute':
            case 'Second':
            case 'DayOfYear':
            case 'DayOfWeek':
            case 'WeekOfYear':
            case 'WeekOfMonth':
                return 'semidiscrete';
        }
        return undefined;
    };
    rangeSelectorItem._getDateTimeScaleType = function (dateTimeGroupInterval) {
        switch (dateTimeGroupInterval) {
            case 'Year':
            case 'QuarterYear':
            case 'MonthYear':
            case 'WeekYear':
            case 'DayMonthYear':
                return 'semidiscrete';
        }
        return null;
    };
    rangeSelectorItem.prototype._getSelectedRangeChangedHandler = function () {
        var that = this;
        return function (e) {
            if (!that._isLocked()) {
                that.selected.fire(that.getName(), _common_1.viewerActions.setMasterFilter, [e.value]);
            }
        };
    };
    rangeSelectorItem.prototype._resizeUnsafe = function () {
        _super.prototype._resizeUnsafe.call(this);
        if ($(this.container).is(':visible')) {
            this.rangeSelectorViewer.render();
        }
    };
    rangeSelectorItem.prototype._getWidget = function () {
        return this.rangeSelectorViewer;
    };
    rangeSelectorItem.prototype._elementCustomColor = function (eventArgs) {
        this.itemElementCustomColor.fire(this.getName(), eventArgs);
    };
    rangeSelectorItem.prototype._hasTimePeriods = function () {
        return true;
    };
    rangeSelectorItem.prototype._isDateTimePeriodSupported = function () {
        return this.options.ViewModel.SupportDateTimePeriods;
    };
    rangeSelectorItem.prototype._getSpecificActionToolbarItems = function () {
        var _this = this;
        if (this.options.ViewModel.SupportDateTimePeriods) {
            if (this.options.ViewModel.DateTimePeriods.length > 0) {
                var dateTimePeriodNames = this.options.ViewModel.DateTimePeriods.map(function (period) { return (period.Name); });
                dateTimePeriodNames.push(_localizer_1.localizer.getString(_localization_ids_1.localizationId.buttonNames.CustomPeriod));
                return [{
                        menu: {
                            name: caption_toolbar_options_1.dashboardToolbarItemNames.dateTimePeriodMenu,
                            items: dateTimePeriodNames,
                            selectedItems: this.timePeriodMenuSelectedIndex !== undefined ? [dateTimePeriodNames[this.timePeriodMenuSelectedIndex]] : [],
                            selectionMode: 'multiple',
                            itemClick: function (itemData, itemElement, index) {
                                var viewModel = _this.options.ViewModel, dateTimePeriods = viewModel.DateTimePeriods;
                                if (index >= 0 && index < dateTimePeriods.length) {
                                    if (_this.timePeriodMenuSelectedIndex !== index) {
                                        _this._raisePredefinedPeriodSelected(index);
                                    }
                                }
                                else {
                                    _this._showCustomTimePeriodDialog();
                                }
                            },
                            type: 'list'
                        },
                        icon: _caption_toolbar_css_classes_1.cssClasses.iconTimePeriods,
                        type: 'menu',
                        hint: _localizer_1.localizer.getString(_localization_ids_1.localizationId.buttonNames.AddTimePeriod),
                    }];
            }
            else {
                return [{
                        name: caption_toolbar_options_1.dashboardToolbarItemNames.customDateTimePeriod,
                        icon: _caption_toolbar_css_classes_1.cssClasses.iconTimePeriods,
                        type: 'button',
                        hint: _localizer_1.localizer.getString(_localization_ids_1.localizationId.buttonNames.AddTimePeriod),
                        click: function (element) {
                            _this._showCustomTimePeriodDialog();
                        }
                    }];
            }
        }
        else
            return [];
    };
    rangeSelectorItem.prototype._raisePredefinedPeriodSelected = function (index) {
        var name = this.options.ViewModel.DateTimePeriods[index].Name;
        this.predefinedRangeChanged && this.predefinedRangeChanged(name);
    };
    rangeSelectorItem.prototype._showCustomTimePeriodDialog = function () {
        this._getCustomTimePeriodDialog().show({
            range: this._getCurrentRange(),
            groupInterval: this.options.ViewModel.Argument.DateTimeGroupInterval,
            isIntYearGroupInterval: this._isIntYearGroupInterval(),
            firstDayOfWeek: this._getFirstDayOfWeek(),
            displayFormat: this._getSliderMarkerFormat()
        });
    };
    rangeSelectorItem.prototype._isIntYearGroupInterval = function () {
        var viewModel = this.options.ViewModel, argument = viewModel.Argument;
        return argument && argument.DateTimeGroupInterval === 'Year' && argument.Type !== 'DateTime';
    };
    return rangeSelectorItem;
}(_base_item_1.baseItem));
exports.rangeSelectorItem = rangeSelectorItem;
var rangeFilterSelection = (function () {
    function rangeFilterSelection(range) {
        this.setMinimum(range.startValue);
        this.setMaximum(range.endValue);
    }
    rangeFilterSelection.prototype.getMaximum = function () {
        return this.maximum;
    };
    rangeFilterSelection.prototype.setMaximum = function (value) {
        this.maximum = value;
    };
    rangeFilterSelection.prototype.getMinimum = function () {
        return this.minimum;
    };
    rangeFilterSelection.prototype.setMinimum = function (value) {
        this.minimum = value;
    };
    return rangeFilterSelection;
}());
exports.rangeFilterSelection = rangeFilterSelection;
