// devexpress-dashboard/designer/conditional-formatting/items/properties-composers/_card-item-format-rule-properties-composer.js
/**
* DevExpress Dashboard (_card-item-format-rule-properties-composer.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var card_item_delta_format_rule_1 = require("../../../../model/format-rules/card-item-delta-format-rule");
var card_item_format_rule_1 = require("../../../../model/format-rules/card-item-format-rule");
var _card_item_delta_format_rule_1 = require("../../../../model/format-rules/metadata/_card-item-delta-format-rule");
var _card_item_format_rule_1 = require("../../../../model/format-rules/metadata/_card-item-format-rule");
var _card_item_format_rule_base_1 = require("../../../../model/format-rules/metadata/_card-item-format-rule-base");
var _card_layout_template_element_1 = require("../../../../model/items/card/metadata/_card-layout-template-element");
var _delta_options_1 = require("../../../../model/items/options/metadata/_delta-options");
var _base_metadata_1 = require("../../../../model/metadata/_base-metadata");
var _form_adapter_editors_1 = require("../../../form-adapter/_form-adapter-editors");
var _display_name_provider_1 = require("../../../_display-name-provider");
var _shared_format_rule_properties_composer_1 = require("./_shared-format-rule-properties-composer");
function getCardItemFormatRuleDataItems(dashboardItem) {
    return dashboardItem.seriesDimensions().concat(dashboardItem.hiddenMeasures());
}
function isDataItemFormatRuleAvaliable(dashboardItem) {
    return getCardItemFormatRuleDataItems(dashboardItem).length > 0;
}
exports.isDataItemFormatRuleAvaliable = isDataItemFormatRuleAvaliable;
function createCardItemFormatRulePropertiesComposer(selectedRuleContainer) {
    return new _shared_format_rule_properties_composer_1.FormatRulePropertiesComposer({
        createDataTypeObservable: _shared_format_rule_properties_composer_1.createDataTypeObservable,
        getCommonFormatRuleProperties: function (formatRule, dashboardItem, dataSourceBrowser) {
            var propertiesInfo = {
                properties: [],
                disabledFilterRules: {},
                dynamicEditorRules: {},
                visibilityFilterRules: {}
            };
            var dataItems = getCardItemFormatRuleDataItems(dashboardItem);
            var dataItemsDisplayText = ko.pureComputed(function () {
                return dataItems.map(function (dataItem) {
                    return {
                        value: dataItem.uniqueName(),
                        displayValueId: _display_name_provider_1.getDataItemDisplayName(dataSourceBrowser, dashboardItem, dataItem)
                    };
                });
            });
            if (!formatRule.dataItemName() && dataItems.length) {
                formatRule.dataItemName(dataItems[0].uniqueName());
            }
            propertiesInfo.properties.push(__assign({}, _card_item_format_rule_1.cardItemformatRuleDataItem, { formAdapterItem: _form_adapter_editors_1.dynamicSelectBoxEditor(dataItemsDisplayText) }));
            return getCommonCardFormatRuleProperties(formatRule, dashboardItem, propertiesInfo, selectedRuleContainer);
        },
        getMiscFormatRuleProperties: function () { return ({ properties: [] }); },
        conditionTypeFilter: exports.cardItemConditionTypeFilter,
        getConditionFormatRuleProperties: getConditionFormatRuleProperties,
    });
}
exports.createCardItemFormatRulePropertiesComposer = createCardItemFormatRulePropertiesComposer;
function cardItemDeltaFormatRuleCards(dashboardItem) {
    return dashboardItem.cards();
}
function isDeltaFormatRuleAvaliable(dashboardItem) {
    return cardItemDeltaFormatRuleCards(dashboardItem).length > 0;
}
exports.isDeltaFormatRuleAvaliable = isDeltaFormatRuleAvaliable;
function createCardItemDeltaFormatRulePropertiesComposer(selectedRuleContainer) {
    return new _shared_format_rule_properties_composer_1.FormatRulePropertiesComposer({
        createDataTypeObservable: _shared_format_rule_properties_composer_1.createDeltaDataTypeObservable,
        getCommonFormatRuleProperties: function (formatRule, dashboardItem, dataSourceBrowser, requestRecalculation) {
            var propertiesInfo = {
                properties: [],
                disabledFilterRules: {},
                dynamicEditorRules: {},
                visibilityFilterRules: {}
            };
            var card = cardItemDeltaFormatRuleCards(dashboardItem);
            var cardsDisplayText = ko.pureComputed(function () {
                return card.map(function (card) {
                    return {
                        value: card._getDataId(),
                        displayValueId: _display_name_provider_1.getDataItemContainerDisplayName(dataSourceBrowser, dashboardItem, card)
                    };
                });
            });
            if (!formatRule.cardId() && cardsDisplayText().length) {
                formatRule.cardId(cardsDisplayText()[0].value);
            }
            propertiesInfo.disabledFilterRules[_card_item_delta_format_rule_1.cardId.propertyName] = function (_) { return !!formatRule.condition(); };
            propertiesInfo.disabledFilterRules[_card_item_delta_format_rule_1.deltaValueType.propertyName] = function (_) { return !!formatRule.condition(); };
            propertiesInfo.properties.push(__assign({}, _card_item_delta_format_rule_1.cardId, { formAdapterItem: _form_adapter_editors_1.dynamicSelectBoxEditor(cardsDisplayText) }));
            propertiesInfo.properties.push(__assign({}, _card_item_delta_format_rule_1.deltaValueType, { simpleFormAdapterItem: 'selectBoxEditor', values: _delta_options_1.deltaValueTypeMap }));
            propertiesInfo.properties.push(__assign({}, _card_item_delta_format_rule_1.deltaValueType, { replacementPropertyName: 'fakeDeltaValueTypeActual', sourceObject: ko.observable('ActualValue'), simpleFormAdapterItem: 'selectBoxEditor', values: { 'ActualValue': 'DashboardStringId.DeltaValueTypeActualValueCaption' } }));
            propertiesInfo.properties.push(__assign({}, _card_item_delta_format_rule_1.deltaValueType, { sourceObject: ko.observable('TargetValue'), replacementPropertyName: 'fakeDeltaValueTypeTarget', simpleFormAdapterItem: 'selectBoxEditor', values: { 'TargetValue': 'DashboardStringId.DeltaValueTypeTargetValueCaption' } }));
            var getActualCard = function () { return dashboardItem.cards().filter(function (card) { return card._getDataId() === formatRule.cardId(); })[0]; };
            var hasActualDataItem = function () { return !!getActualCard().actualValue(); };
            var hasTargetDataItem = function () { return !!getActualCard().targetValue(); };
            requestRecalculation.add(function (_) {
                if (dashboardItem.formatRules.indexOf(formatRule) === -1) {
                    if (hasActualDataItem() && !hasTargetDataItem())
                        formatRule.deltaValueType('ActualValue');
                    else if (!hasActualDataItem() && hasTargetDataItem()) {
                        formatRule.deltaValueType('TargetValue');
                    }
                }
            });
            propertiesInfo.visibilityFilterRules[_card_item_delta_format_rule_1.deltaValueType.propertyName] = function (_) { return hasActualDataItem() && hasTargetDataItem(); };
            propertiesInfo.visibilityFilterRules['fakeDeltaValueTypeActual'] = function (_) { return hasActualDataItem() && !hasTargetDataItem(); };
            propertiesInfo.visibilityFilterRules['fakeDeltaValueTypeTarget'] = function (_) { return !hasActualDataItem() && hasTargetDataItem(); };
            propertiesInfo.disabledFilterRules['fakeDeltaValueTypeActual'] = function (_) { return true; };
            propertiesInfo.disabledFilterRules['fakeDeltaValueTypeTarget'] = function (_) { return true; };
            return getCommonCardFormatRuleProperties(formatRule, dashboardItem, propertiesInfo, selectedRuleContainer);
        },
        getConditionFormatRuleProperties: getConditionFormatRuleProperties,
        getMiscFormatRuleProperties: function () { return ({ properties: [] }); },
        conditionTypeFilter: exports.cardItemConditionTypeFilter
    });
}
exports.createCardItemDeltaFormatRulePropertiesComposer = createCardItemDeltaFormatRulePropertiesComposer;
function getConditionFormatRuleProperties(formatRule, dashboardItem) {
    return {
        properties: [__assign({}, _card_item_format_rule_base_1.layoutItemApplyTo, { simpleFormAdapterItem: 'selectBoxEditor', values: _card_layout_template_element_1.сardFormatRuleLayoutElementValuesMap })],
        disabledFilterRules: {},
        dynamicEditorRules: {},
        visibilityFilterRules: {}
    };
}
exports.cardItemConditionTypeFilter = function (conditionTypePropertyName) {
    return ['conditionBar', 'conditionColorRangeBar', 'conditionGradientRangeBar'].indexOf(conditionTypePropertyName) === -1;
};
function getCommonCardFormatRuleProperties(formatRule, dashboardItem, calculatedByProperties, selectedRuleContainer) {
    var _a;
    var propertiesInfo = {
        properties: [],
        disabledFilterRules: {},
        dynamicEditorRules: {},
        visibilityFilterRules: {}
    };
    var selectedRuleType = ko.observable(formatRule.itemType());
    selectedRuleType.subscribe(function (newSelectedContainer) {
        if (newSelectedContainer === 'CardItemFormatRule') {
            selectedRuleContainer(new card_item_format_rule_1.CardItemFormatRule());
        }
        else if (newSelectedContainer === 'CardItemDeltaFormatRule') {
            selectedRuleContainer(new card_item_delta_format_rule_1.CardItemDeltaFormatRule());
        }
    });
    propertiesInfo.properties.push(__assign({}, _base_metadata_1.itemType, { sourceObject: selectedRuleType, replacementPropertyName: 'selectedContainer', displayName: 'DashboardWebStringId.ConditionalFormatting.CalculatedBy', formAdapterItem: _form_adapter_editors_1.buttonGroupEditor([
            {
                displayValueId: 'DashboardWebStringId.ConditionalFormatting.CardRuleTypeCard',
                value: 'CardItemDeltaFormatRule'
            },
            {
                displayValueId: 'DashboardWebStringId.ConditionalFormatting.CardRuleTypeDataItem',
                value: 'CardItemFormatRule'
            }
        ]) }));
    propertiesInfo.disabledFilterRules['selectedContainer'] = function (m) { return !!formatRule.condition() || !isDeltaFormatRuleAvaliable(dashboardItem) || !isDataItemFormatRuleAvaliable(dashboardItem); };
    (_a = propertiesInfo.properties).push.apply(_a, calculatedByProperties.properties);
    propertiesInfo.disabledFilterRules = __assign({}, propertiesInfo.disabledFilterRules, calculatedByProperties.disabledFilterRules);
    propertiesInfo.dynamicEditorRules = __assign({}, propertiesInfo.dynamicEditorRules, calculatedByProperties.dynamicEditorRules);
    propertiesInfo.visibilityFilterRules = __assign({}, propertiesInfo.visibilityFilterRules, calculatedByProperties.visibilityFilterRules);
    return propertiesInfo;
}
