// devexpress-dashboard/viewer-parts/widgets/caption-toolbar/caption-toolbar-options.js
/**
* DevExpress Dashboard (caption-toolbar-options.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dashboardToolbarItemNames = {
    titleFilterText: 'filter-text',
    titleFilterIcon: 'filter-icon',
    exportMenu: 'export-menu',
    dashboardTitleImage: 'dashboard-title-image',
    dashboardTitle: 'dashboard-title',
    parameters: 'parameters',
    itemCaption: 'item-caption',
    limitVisibleData: 'limit-visible-data',
    clearMasterFilter: 'clear-master-filter',
    clearSelection: 'clear-selection',
    drillUp: 'drill-up',
    values: 'values',
    multiselection: 'multiselection',
    initialExtent: 'initial-extent',
    dateTimePeriodMenu: 'date-time-period-menu',
    customDateTimePeriod: 'custom-date-time-period',
    dashboardList: 'dashboard-list',
    dashboardFilters: 'dashboard-filters',
    backButton: 'back-button',
    itemFilters: 'item-filters',
    restoreItem: 'restore-item',
    maximizeItem: 'maximize-down-item',
    dataInspector: 'data-inspector'
};
