// devexpress-dashboard/designer/items/surfaces/_custom-item-surface.js
/**
* DevExpress Dashboard (_custom-item-surface.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var string_1 = require("devextreme/core/utils/string");
var _default_1 = require("../../../data/localization/_default");
var custom_item_1 = require("../../../model/items/custom-item/custom-item");
var accordion_tab_options_1 = require("../../accordion-tab-options");
var _object_properties_wrapper_1 = require("../../form-adapter/_object-properties-wrapper");
var _accordion_tab_1 = require("../../properties-controller/_accordion-tab");
var _base_properties_composer_1 = require("../properties-composers/_base-properties-composer");
var _shared_composers_1 = require("../properties-composers/_shared-composers");
var _data_item_collection_surface_1 = require("../sections/_data-item-collection-surface");
var _single_data_item_surface_1 = require("../sections/_single-data-item-surface");
var _section_descriptors_1 = require("../_section-descriptors");
var _base_item_surface_1 = require("./_base-item-surface");
var _custom_item_properties_composer_helper_1 = require("./_custom-item-properties-composer-helper");
var CustomItemPropertiesComposer = (function (_super) {
    __extends(CustomItemPropertiesComposer, _super);
    function CustomItemPropertiesComposer(customizeHandler) {
        return _super.call(this, customizeHandler) || this;
    }
    CustomItemPropertiesComposer.prototype._mapLegacyPropertiesToTabs = function (model) {
        var propertiesByTabs = model.legacyCustomProperties && model.legacyCustomProperties.reduce(function (acc, property) {
            var tabName = property.sectionName || _accordion_tab_1.KnownTabs.Common;
            acc[tabName] = acc[tabName] || [];
            acc[tabName].push(_custom_item_properties_composer_helper_1.CustomItemPropertiesComposerHelper.convertToDashboardSerializationInfo(property));
            return acc;
        }, {});
        var specificProperties = propertiesByTabs[_accordion_tab_1.KnownTabs.Common] ? [
            {
                container: { propertyName: 'customProperties' },
                properties: propertiesByTabs[_accordion_tab_1.KnownTabs.Common] || []
            }
        ] : [];
        return [
            new _accordion_tab_1.AccordionTab(_accordion_tab_1.KnownTabs.Common, 'DashboardWebStringId.AccordionTab.Common', _shared_composers_1.SharedComposers.getCommonWrapper(model, specificProperties))
        ].concat(Object.keys(propertiesByTabs)
            .filter(function (tabName) { return tabName !== _accordion_tab_1.KnownTabs.Common; })
            .map(function (tabName) { return new _accordion_tab_1.AccordionTab(tabName, tabName, new _object_properties_wrapper_1.ObjectPropertiesWrapper({
            model: model.customProperties,
            properties: propertiesByTabs[tabName]
        })); }));
    };
    CustomItemPropertiesComposer.prototype._composeTabsCore = function (model) {
        if (model._optionsPanelSections) {
            return [
                new _accordion_tab_1.AccordionTab(_accordion_tab_1.KnownTabs.Common, 'DashboardWebStringId.AccordionTab.Common', _shared_composers_1.SharedComposers.getCommonWrapper(model))
            ].concat(model._optionsPanelSections.reduce(function (tabs, section) {
                accordion_tab_options_1._customizeTabs(tabs, section, model);
                return tabs;
            }, []));
        }
        else {
            return this._mapLegacyPropertiesToTabs(model);
        }
    };
    return CustomItemPropertiesComposer;
}(_base_properties_composer_1.DashboardItemPropertiesComposer));
exports.CustomItemPropertiesComposer = CustomItemPropertiesComposer;
var CustomItemSurface = (function (_super) {
    __extends(CustomItemSurface, _super);
    function CustomItemSurface(dashboardItem, dashboardModel, dataSourceBrowser) {
        return _super.call(this, dashboardItem, dashboardModel, dataSourceBrowser) || this;
    }
    Object.defineProperty(CustomItemSurface.prototype, "showDefaultSections", {
        get: function () {
            return this.dashboardItem.customBindings.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    CustomItemSurface.prototype.fillSections = function () {
        var _this = this;
        var format = function (localizationId, arg1) { return string_1.format(_default_1.getLocalizationById(localizationId), _default_1.getLocalizationById(arg1)); };
        this.dashboardItem.customBindings.forEach(function (binding) {
            var sectionInfo = {
                title: binding.displayName,
                bindingProperty: {
                    propertyName: custom_item_1.CustomItem._getBindingLinkName(binding.propertyName),
                    dataItemType: binding.dataItemType,
                    emptyPlaceholder: binding.emptyPlaceholder || format('DashboardWebStringId.Binding.DefaultEmptyPlaceholder', binding.displayName),
                    selectedPlaceholder: binding.selectedPlaceholder || format('DashboardWebStringId.Binding.DefaultSelectedPlaceholder', binding.displayName)
                }
            };
            if (binding.array) {
                _this.dataSections.push(new _data_item_collection_surface_1.DataItemCollectionSurface(_this, sectionInfo));
            }
            else {
                _this.dataSections.push(new _single_data_item_surface_1.SingleDataItemSurface({
                    itemSurface: _this,
                    sectionInfo: sectionInfo,
                    fieldConstraint: function (dataField) {
                        return !binding.constraints || !binding.constraints.allowedTypes || binding.constraints.allowedTypes.indexOf(dataField.fieldType()) >= 0;
                    }
                }));
            }
        });
    };
    CustomItemSurface.prototype.getPropertiesComposer = function () {
        return new CustomItemPropertiesComposer(this._dashboardItemCustomization);
    };
    return CustomItemSurface;
}(_base_item_surface_1.DataDashboardItemSurface));
exports.CustomItemSurface = CustomItemSurface;
_section_descriptors_1.surfaceItemsFactory.register('CustomItem', CustomItemSurface);
