// devexpress-dashboard/designer/color-scheme-editor-extension/internal/_color-picker-model.js
/**
* DevExpress Dashboard (_color-picker-model.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var _default_1 = require("../../../data/localization/_default");
var color_1 = require("../../../model/color");
var color_scheme_entry_1 = require("../../../model/colorization/color-scheme-entry");
var ColorPickerModel = (function () {
    function ColorPickerModel(colorSchemeModel, colorPalette) {
        var _this = this;
        this.colorSchemeModel = colorSchemeModel;
        this.colorPalette = colorPalette;
        this.target = ko.observable();
        this.visible = ko.observable();
        this.colorCss = ko.observable('');
        this.confirm = function () {
            _this.entry.paletteIndex(null);
            _this.entry.color(color_1.Color.fromRgbaString(_this.colorCss()));
            if (_this.entry instanceof color_scheme_entry_1.AutoColorSchemeEntry) {
                var newEntry = _this.entry.clone();
                _this.colorSchemeModel.updateEntry(_this.entry, newEntry);
            }
            _this.visible(false);
        };
        this.buttonItems = ko.observableArray([
            { toolbar: 'bottom', location: 'after', widget: 'dxButton', options: { text: _default_1.getLocalizationById('DashboardWebStringId.ButtonConfirm'), onClick: this.confirm } },
            { toolbar: 'bottom', location: 'after', widget: 'dxButton', options: { text: _default_1.getLocalizationById('DashboardStringId.ButtonCancel'), onClick: function () { return _this.visible(false); } } }
        ]);
    }
    ColorPickerModel.prototype.init = function (entry, target) {
        this.entry = entry;
        var color = entry.paletteIndex() === null ? entry.color() : this.colorPalette()[entry.paletteIndex()];
        this.colorCss(color ? color.css : '');
        this.target(target);
        this.visible(true);
    };
    return ColorPickerModel;
}());
exports.ColorPickerModel = ColorPickerModel;
