// devexpress-dashboard/designer/parameter-settings/parameters-editor-extension.js
/**
* DevExpress Dashboard (parameters-editor-extension.js)
* Version:  20.1.6
* Build date: Jul 17, 2020
* Copyright (c) 2012 - 2020 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ko = require("knockout");
var control_options_1 = require("../../common/control-options");
var toolbox_items_1 = require("../toolbox-extension/toolbox-items");
var _parameter_list_editor_viewmodel_1 = require("./_parameter-list-editor-viewmodel");
var name = 'dashboard-parameter-editor';
var DashboardParameterEditorExtension = (function () {
    function DashboardParameterEditorExtension(dashboardControl) {
        this.dashboardControl = dashboardControl;
        this.name = name;
        this.dashboard = ko.computed(function () { return dashboardControl.dashboard(); });
        this._viewModel = new _parameter_list_editor_viewmodel_1.ParameterListEditorViewModel(this.dashboard, function () { return dashboardControl._dataSourceBrowser; });
        this._menuItem = new toolbox_items_1.DashboardMenuItem(this.name, 'DashboardWebStringId.DashboardParameters', 240, 80);
        this._menuItem.template = 'dx-dashboard-form-parameters-settings';
        this._menuItem.data = this._viewModel;
        this._menuItem.disabled = ko.computed(function () { return !dashboardControl.dashboard(); });
    }
    DashboardParameterEditorExtension.prototype.start = function () {
        this._viewModel.initialize();
        var toolboxExtension = this.dashboardControl.findExtension('toolbox');
        if (toolboxExtension) {
            toolboxExtension.menuItems.push(this._menuItem);
        }
    };
    DashboardParameterEditorExtension.prototype.stop = function () {
        var toolboxExtension = this.dashboardControl.findExtension('toolbox');
        if (toolboxExtension) {
            toolboxExtension.menuItems.remove(this._menuItem);
        }
        this._viewModel.dispose();
    };
    return DashboardParameterEditorExtension;
}());
exports.DashboardParameterEditorExtension = DashboardParameterEditorExtension;
control_options_1.designerExtensions[name] = function (dashboardControl, options) { return new DashboardParameterEditorExtension(dashboardControl); };
